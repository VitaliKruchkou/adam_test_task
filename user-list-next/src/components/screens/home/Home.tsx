import React, { ChangeEvent, FC, useEffect, useState } from 'react'

import Layout from '@/components/layout/Layout'
import { User, UserData } from '@/interfaces/user.interface'
import UserItem from '@/components/pages/user/UserItem'
import { getFullName } from '@/utils'

import { useStyles } from './Home.style'
import HomeForm from './HomeForm'

const Home: FC<UserData> = ({ users }) => {
  const [searchAndSortedUsers, setSearchAndSortedUsers] = useState<User[]>(users)
  const [searchValue, setSearchValue] = useState<string>('')

  const styles = useStyles()

  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    setSearchValue(e.target.value)
  }

  const handleSelectChange = (e: ChangeEvent<HTMLSelectElement>) => {
    const sortDirection = e.target.value as keyof Omit<User, 'id' | 'additionalInformation'>

    const sortedUsersByValue = [...users].sort((user1: User, user2: User) =>
      user1[sortDirection!]?.localeCompare(user2[sortDirection!]),
    )

    setSearchAndSortedUsers(sortedUsersByValue)
  }

  useEffect(() => {
    const filteredUsersData = users.filter((user: User) =>
      getFullName(user).toLowerCase().includes(searchValue.toLowerCase()),
    )

    setSearchAndSortedUsers(filteredUsersData)
  }, [searchValue])

  return (
    <Layout title="Home" description="Users list">
      <HomeForm
        searchValue={searchValue}
        handleInputChange={handleInputChange}
        handleSelectChange={handleSelectChange}
      />
      <div className={styles.container} aria-label="cost-container">
        {searchAndSortedUsers?.length > 0 ? (
          searchAndSortedUsers.map((user: User) => <UserItem key={user.id} user={user} />)
        ) : (
          <div>User not found</div>
        )}
      </div>
    </Layout>
  )
}

export default Home
