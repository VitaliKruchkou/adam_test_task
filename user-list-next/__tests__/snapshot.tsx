/**
 * @jest-environment jsdom
 */
import Home from '@/components/screens/home/Home'
import { render } from '@testing-library/react'

jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}))

it('renders homepage unchanged', () => {
  const { container } = render(<Home users={[]} />)
  expect(container).toMatchSnapshot()
})
