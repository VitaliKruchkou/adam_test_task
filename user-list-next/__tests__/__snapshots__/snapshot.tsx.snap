// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders homepage unchanged 1`] = `
<div>
  <header
    class="css-urcdts"
  >
    <a
      class="css-18ql7pt"
      href="/"
    >
      Home
    </a>
    <a
      class="css-urcdts"
      href="/about"
    >
      About page
    </a>
  </header>
  <main>
    <input
      aria-label="cost-input"
      placeholder="Search ..."
      type="text"
      value=""
    />
    <select
      aria-label="cost-select"
    >
      <option
        value=""
      >
        Select from the list
      </option>
      <option
        value="dateOfBirth"
      >
        Date of Birthday
      </option>
    </select>
    <div
      aria-label="cost-container"
      class="css-1s806mh"
    >
      <div>
        User not found
      </div>
    </div>
  </main>
</div>
`;
